js/personalize.storage.js

```diff
@@ -0,0 +1,191 @@
+(function (Drupal) {
+  Drupal.personalizeStorage = (function() {
+
+    var personalizeStorageKey = 'personalize::storage::keys';
+
+     /**
+     * Determine if the current browser supports web storage.
+     */
+    function _supportsLocalStorage() {
+      if (this.supportsHtmlLocalStorage != undefined) {
+        return this.supportsHtmlLocalStorage;
+      }
+      try {
+        this.supportsHtmlLocalStorage = 'localStorage' in window && window['localStorage'] !== null;
+      } catch (e) {
+        this.supportsHtmlLocalStorage = false;
+      }
+      return this.supportsHtmlLocalStorage;
+    }
+
+    /**
+     * Determine the kind of storage to use based on session type requested.
+     *
+     * @param session
+     *   True if session storage, false if local storage.  Defaults true.
+     */
+    function _getStore(session) {
+      session = typeof(session) == 'undefined' ? true : session;
+      return session ? sessionStorage : localStorage;
+    }
+
+    /**
+     * Gets the listing of keys and the order in which they were added.
+     *
+     * @param key
+     *   The key of the item saved to storage
+     * @param session
+     *   True if session storage, false if local storage.  Defaults true.
+     */
+    function _getTrackedKeys(session) {
+      var store = _getStore(session);
+      var keys = store.getItem(personalizeStorageKey);
+      if (keys) {
+        keys = JSON.parse(keys);
+      } else {
+        keys = [];
+      }
+      return keys;
+    }
+
+    /**
+     * Update the listing of keys and the order in which they were added.
+     *
+     * @param key
+     *   The key of the item saved to storage
+     * @param session
+     *   True if session storage, false if local storage.  Defaults true.
+     */
+    function _trackKey(key, session) {
+      var store = _getStore(session);
+      var keys = _getTrackedKeys(session);
+      keys.push(key);
+      store.setItem(personalizeStorageKey, JSON.stringify(keys));
+    }
+
+    /**
+     * Prunes the oldest key(s) from storage.
+     *
+     * @param session
+     *   True if session storage, false if local storage.  Defaults true.
+     * @param numEntries
+     *   The number of entries to remove.  Default = 10.
+     */
+    function _pruneOldest(session, numEntries) {
+      numEntries = numEntries || 10;
+      var keys = _getTrackedKeys(session);
+      var totalKeys = keys.length;
+      var until = totalKeys > numEntries ? totalKeys - numEntries : 0;
+
+      for (var i=totalKeys; i>=until; i--) {
+        key = keys.pop();
+        this.removeItem(key, session);
+      }
+    }
+
+    return {
```

Comment:
These public methods need qunit tests for unit testing.  I'd be interested in seeing tests to ensure basic read/write as well as pruning and tests (if possible) to ensure things work when no local storage is available.